sdk/eventhub/event-hubs/src/receiver.ts

```diff
@@ -232,61 +262,159 @@ export class EventHubConsumer {
-        // This can result in the receiver link having pending credits if a receiveBatch() call asking for m events returned n events where n < m.
-        // Since Event Hubs doesnt support the drain feature yet, this can result in the receiver link receiving events when the user is not expecting it to.
-        // Hence closing the link when result.length < maxMessageCount
-        await this._batchingReceiver.close();
-      }
-    } catch (err) {
-      log.error(
-        "[%s] Receiver '%s', an error occurred while receiving %d messages for %d max time:\n %O",
-        this._context.connectionId,
-        this._batchingReceiver.name,
-        maxMessageCount,
-        maxWaitTimeInSeconds,
-        err
-      );
-      throw err;
-    }
-    return result;
+    // store events across multiple retries
+    const receivedEvents: ReceivedEventData[] = [];
+
+    const retrieveEvents = (): Promise<ReceivedEventData[]> => {
+      return new Promise(async (resolve, reject) => {
+        // if this consumer was closed, _baseConsumer might be undefined.
+        // resolve the operation's promise with the events collected thus far in case
+        // the promise hasn't already been resolved.
+        if (!this._baseConsumer) {
+          return resolve(receivedEvents);
+        }
+
+        let timer: any;
+        const logOnAbort = (): void => {
+          const baseConsumer = this._baseConsumer;
+          const name = baseConsumer && baseConsumer.name;
+          const address = baseConsumer && baseConsumer.address;
+          const desc: string =
+            `[${this._context.connectionId}] The request operation on the Receiver "${name}" with ` +
+            `address "${address}" has been cancelled by the user.`;
+          log.error(desc);
+        };
+
+        const rejectOnAbort = async (): Promise<void> => {
+          logOnAbort();
+          try {
+            await this.close();
+          } finally {
+            return reject(new AbortError("The receive operation has been cancelled by the user."));
+          }
+        };
+
+        // operation has been cancelled, so exit immediately
+        if (abortSignal && abortSignal.aborted) {
+          return await rejectOnAbort();
+        }
+
+        const onAbort = (): void => {
+          clearTimeout(timer);
+          rejectOnAbort();
+        };
+
+        // updates the prefetch count so that the baseConsumer adds
+        // the correct number of credits to receive the same number of events.
+        const prefetchCount = Math.max(maxMessageCount - receivedEvents.length, 0);
+        if (prefetchCount === 0) {
+          return resolve(receivedEvents);
+        }
+
+        log.batching(
+          "[%s] Receiver '%s', setting the prefetch count to %d.",
+          this._context.connectionId,
+          this._baseConsumer && this._baseConsumer.name,
+          prefetchCount
+        );
+        this._baseConsumer.prefetchCount = prefetchCount;
```

Comment:
Prefetch count is being set before `baseconsumer.intialize()` is run. This would result in the underlying amqp receiver link being created with a `credit_window` equal to the prefetch count.

Setting credit_window to a non zero value means that we will be opting in to automatic credit management i.e rhea will keep adding credits automatically to maintain this window. While we may want this for streaming receiver, we definitely do not want this for batching receiver.

Reply:
Yes, I shouldn’t be setting this until after handlers are registered and the link has been initialized. I’ll be changing this bit as well as more directly handling the credits in this class.